tty/pc-tty/ttypc_kbd.c

```diff
@@ -596,34 +656,92 @@ int ttypc_kbd_init(ttypc_t *ttypc)
 	/* Configure typematic */
 	do {
 		/* Send set typematic rate/delay command */
-		if (ttypc_kbd_write(ttypc, 0xf3) < 0)
+		err = ttypc_ps2_write(ttypc, 0xf3);
+		if (err < 0) {
 			break;
+		}
 
 		/* 250 ms / 30.0 reports/sec */
-		if (ttypc_kbd_write(ttypc, 0) < 0)
+		err = ttypc_ps2_write(ttypc, 0);
+		if (err < 0) {
 			break;
+		}
 	} while (0);
 
-	if ((err = mutexCreate(&ttypc->klock)) < 0)
-		return err;
+#if PC_TTY_CREATE_PS2_VDEVS
+	oid_t oid;
 
-	if ((err = condCreate(&ttypc->kcond)) < 0) {
-		resourceDestroy(ttypc->klock);
+	/* Wait for the filesystem */
+	while (lookup("/", NULL, &oid) < 0) {
+		usleep(10000);
+	}
+
+	/* Create virtual keyboard device */
+	err = portCreate(&ttypc->kport);
+	if (err < 0) {
+		(void)fprintf(stderr, "pc-tty: failed to create keyboard port\n");
 		return err;
 	}
 
-	/* Attach interrupt */
-	if ((err = interrupt((ttypc->kirq = 1), ttypc_kbd_interrupt, ttypc, ttypc->kcond, &ttypc->kinth)) < 0) {
-		resourceDestroy(ttypc->klock);
-		resourceDestroy(ttypc->kcond);
+	oid.port = ttypc->kport;
+	oid.id = 0;
+
+	err = create_dev(&oid, "/dev/kbd");
+	if (err < 0) {
+		(void)fprintf(stderr, "pc-tty: failed to register kbd device\n");
 		return err;
 	}
+#endif
+
+	do {
+		err = mutexCreate(&ttypc->kmlock);
+		if (err < 0) {
+			break;
+		}
+
+		err = condCreate(&ttypc->kmcond);
+		if (err < 0) {
+			break;
+		}
+
+#if PC_TTY_CREATE_PS2_VDEVS
+		err = event_queue_init(&ttypc->keq);
+		if (err < 0) {
+			break;
+		}
+#endif
+
+		/* Attach KIRQ1 (kbd event) interrupt handle */
+		err = interrupt((ttypc->kirq = 1), _ttypc_kbd_interrupt, ttypc, ttypc->kmcond, &ttypc->kinth);
+		if (err < 0) {
+			break;
+		}
+
+		/* Initialize mouse */
+		err = ttypc_mouse_init(ttypc);
```

Comment:
Does it have to be inside `ttypc_kbd_init()` or can it be moved outside? I assume mouse and keyboard are intertwined a bit and it can be cumbersome to separate these functions

Reply:
They share the data ports so they are heavily intertwined, but it *may* be possible to separate. Here I wanted to ensure that the controller is fully initialized before attempting to do any reads in the `ttypc_ps2_ctlthr` thread